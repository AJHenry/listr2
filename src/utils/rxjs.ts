export { Observable } from 'rxjs'
export { Subject } from 'rxjs'
