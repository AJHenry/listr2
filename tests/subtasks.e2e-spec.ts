import delay from 'delay'
import { mockProcessExit, mockProcessStderr, mockProcessStdout } from 'jest-mock-process'
import { Observable } from 'rxjs'

import { Listr } from '@root/index'

describe('show output from task', () => {
  let mockExit: jest.SpyInstance<never, [number?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStdout: jest.SpyInstance<boolean, [string, string?, Function?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStderr: jest.SpyInstance<boolean, [string, string?, Function?]>

  beforeEach(async () => {
    mockExit = mockProcessExit()
    mockStdout = mockProcessStdout()
    mockStderr = mockProcessStderr()
  })

  afterEach(async () => {
    mockExit.mockRestore()
    mockStdout.mockRestore()
    mockStderr.mockRestore()
    jest.clearAllMocks()
  })

  // test-e2e-JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR
  it('should create a set of subtasks', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx, task): Listr => task.newListr([
            {
              title: 'This is a subtask.',
              task: async (): Promise<void> => {
                await delay(30)
              }
            },

            {
              title: 'This is a subtask.',
              task: async (): Promise<void> => {
                await delay(30)
              }
            }
          ])
        }
      ],
      {
        concurrent: false, nonTTYRenderer: 'default', nonTTYRendererOptions: { lazy: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('test-e2e-JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('test-e2e-JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-err')
    expect(mockExit.mock.calls).toMatchSnapshot('test-e2e-JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-exit')
  })

  // test-e2e-3ygUxrTkCDLDfDrXzF1ocWR6626jaL0kA
  it('should be able to change concurrency setting in subtask, second subtask should finish first', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx, task): Listr => task.newListr([
            {
              title: 'This is a subtask.',
              task: async (): Promise<void> => {
                await delay(30)
              }
            },

            {
              title: 'This is an another subtask.',
              task: async (): Promise<void> => {
                await delay(20)
              }
            }
          ], { concurrent: true })
        }

      ],
      {
        concurrent: false, nonTTYRenderer: 'default', nonTTYRendererOptions: { lazy: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('test-e2e-3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('test-e2e-3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-err')
    expect(mockExit.mock.calls).toMatchSnapshot('test-e2e-3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-exit')
  })

  // test-e2e-8TNRrm8bI9ndz4jrhYC492luGNhtMTmZ
  it('should be able to change collapse settings', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx, task): Listr => task.newListr([
            {
              title: 'This is a subtask.',
              task: async (): Promise<void> => {
                await delay(30)
              }
            },

            {
              title: 'This is an another subtask.',
              task: async (): Promise<void> => {
                await delay(20)
              }
            }
          ])
        },

        {
          title: 'This task will execute.',
          task: (ctx, task): Listr => task.newListr([
            {
              title: 'This is a subtask.',
              task: async (): Promise<void> => {
                await delay(30)
              }
            },

            {
              title: 'This is an another subtask.',
              task: async (): Promise<void> => {
                await delay(20)
              }
            }
          ], { rendererOptions: { collapse: false } })
        }

      ],
      {
        concurrent: false, nonTTYRenderer: 'default', nonTTYRendererOptions: { lazy: true, collapse: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('test-e2e-8TNRrm8bI9ndz4jrhYC492luGNhtMTmZ-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('test-e2e-8TNRrm8bI9ndz4jrhYC492luGNhtMTmZ-err')
    expect(mockExit.mock.calls).toMatchSnapshot('test-e2e-8TNRrm8bI9ndz4jrhYC492luGNhtMTmZ-exit')
  })

})
