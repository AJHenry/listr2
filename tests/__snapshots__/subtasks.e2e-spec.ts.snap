// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show output from task should be able to change collapse settings: test-e2e-8TNRrm8bI9ndz4jrhYC492luGNhtMTmZ-err 1`] = `Array []`;

exports[`show output from task should be able to change collapse settings: test-e2e-8TNRrm8bI9ndz4jrhYC492luGNhtMTmZ-exit 1`] = `Array []`;

exports[`show output from task should be able to change collapse settings: test-e2e-8TNRrm8bI9ndz4jrhYC492luGNhtMTmZ-out 1`] = `
Array [
  Array [
    "[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[G[2m‚óº[22m This task will execute.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.

[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [2m‚óº[22m This is a subtask.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [2m‚óº[22m This is a subtask.
  [2m‚óº[22m This is an another subtask.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [33m‚ùØ[39m This is a subtask.
  [2m‚óº[22m This is an another subtask.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [2m‚óº[22m This is an another subtask.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [33m‚ùØ[39m This is an another subtask.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [32m‚úî[39m This is an another subtask.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[33m‚ùØ[39m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[33m‚ùØ[39m This task will execute.
  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[33m‚ùØ[39m This task will execute.
  [2m‚óº[22m This is a subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[33m‚ùØ[39m This task will execute.
  [2m‚óº[22m This is a subtask.
  [2m‚óº[22m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[33m‚ùØ[39m This task will execute.
  [33m‚ùØ[39m This is a subtask.
  [2m‚óº[22m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [2m‚óº[22m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [33m‚ùØ[39m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [32m‚úî[39m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[32m‚úî[39m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.
[32m‚úî[39m This task will execute. 
",
  ],
]
`;

exports[`show output from task should be able to change concurrency setting in subtask, second subtask should finish first: test-e2e-3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-err 1`] = `Array []`;

exports[`show output from task should be able to change concurrency setting in subtask, second subtask should finish first: test-e2e-3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-exit 1`] = `Array []`;

exports[`show output from task should be able to change concurrency setting in subtask, second subtask should finish first: test-e2e-3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-out 1`] = `
Array [
  Array [
    "[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[G[33m‚ùØ[39m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [2m‚óº[22m This is a subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [2m‚óº[22m This is a subtask.
  [2m‚óº[22m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [33m‚ùØ[39m This is a subtask.
  [2m‚óº[22m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [33m‚ùØ[39m This is a subtask.
  [33m‚ùØ[39m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [33m‚ùØ[39m This is a subtask.
  [32m‚úî[39m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [32m‚úî[39m This is an another subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[G[32m‚úî[39m This task will execute. 
",
  ],
]
`;

exports[`show output from task should create a set of subtasks: test-e2e-JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-err 1`] = `Array []`;

exports[`show output from task should create a set of subtasks: test-e2e-JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-exit 1`] = `Array []`;

exports[`show output from task should create a set of subtasks: test-e2e-JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-out 1`] = `
Array [
  Array [
    "[2m‚óº[22m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[G[33m‚ùØ[39m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [2m‚óº[22m This is a subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [2m‚óº[22m This is a subtask.
  [2m‚óº[22m This is a subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [33m‚ùØ[39m This is a subtask.
  [2m‚óº[22m This is a subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [2m‚óº[22m This is a subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [33m‚ùØ[39m This is a subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[33m‚ùØ[39m This task will execute.
  [32m‚úî[39m This is a subtask.
  [32m‚úî[39m This is a subtask.  
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[1A[2K[G[32m‚úî[39m This task will execute.  
",
  ],
  Array [
    "[2K[1A[2K[G[32m‚úî[39m This task will execute. 
",
  ],
]
`;
